toolchain/parser/parser_handle_index_expression.cpp

```diff
@@ -19,6 +19,13 @@ auto ParserHandleIndexExpression(ParserContext& context) -> void {
 
 auto ParserHandleIndexExpressionFinish(ParserContext& context) -> void {
   auto state = context.PopState();
+  if (!context.PositionIs(TokenKind::CloseSquareBracket)) {
+    CARBON_DIAGNOSTIC(UnexpectedTokenInIndex, Error,
+                      "Unexpected token in index expression");
+    context.emitter().Emit(*context.position(), UnexpectedTokenInIndex);
+    context.ReturnErrorOnState();
```

Comment:
I think we should mark the error on the `IndexExpression` parse node, not the enclosing node:
```suggestion
    state.has_error = true;
```